Fix bugs

melanieseltzer opened the pull request:
**Summary:**

The Application Menu was bugging out and 'Current Project' menu item was not showing upon selecting project, and only showed after refreshing the app. 

After some tinkering around I figured out that `componentWillReceiveProps()` was causing it to be null on first try, and refreshing was the only way to trigger it to show. So I solved it by switching to `componentDidUpdate()` instead, per [the docs](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops) (seems like they're recommending to not use it anymore) and it seems to be working much better. 

We might want to think about switching other instances of `componentWillReceiveProps()` as well?

Also implemented the onboarding status check per our Gitter discussion. So now the sidebar instructions only show if onboardingStatus is 'introducing-sidebar'. And if 'done' then we just select the project and don't show the instructions again.

joshwcomeau commented on src/components/CreateNewProjectWizard/CreateNewProjectWizard.js:
```diff
@@ -103,10 +105,15 @@ class CreateNewProjectWizard extends PureComponent<Props, State> {
+    const { onboardingStatus } = this.props;
+
+    // Get onboardingStatus to check if sidebar instructions should display
+    const onboardingCompleted = onboardingStatus === 'done';
```
I would actually do this check in `mapStateToProps`. I like the idea of giving components as little information as they need, and `CreateNewProjectWizard` doesn't actually need to know the exact status, it just needs to know if it's done or not.

So I'd make the prop itself `isOnboardingCompleted` and then just pass it through this component to the action.

Another nice touch we could do is to add a selector function to `onboarding-status.reducer.js`, like:

```js
export const getOnboardingCompleted = (state: GlobalState) => 
  getOnboardingStatus(state) === 'done';
```

Then in this component you would just do:

```js
const mapStateToProps = state => ({
  isOnboardingCompleted: getOnboardingCompleted(state),
})
```

The reason this is nice is that the `onboarding-status.reducer` "owns" the bit of information about what the "done" state is. Let's say in the future we want to change how onboarding statuses work, we'd only have to change that one file, rather than finding all the places where it checks for raw values in components.

melanieseltzer replied:
Oh man, this is why I love reviews, such valuable info. Never thought to do this in mapStateToProps and in the reducer but it's so much cleaner. Plus the reducer has the matching `getOnboardingStatus` so it makes total sense to keep this done check in there as well.

joshwcomeau commented on src/actions/index.js:
```diff
@@ -62,9 +62,10 @@ export const RESET_ALL_STATE = 'RESET_ALL_STATE';
 // Action Creators
 //
-export const addProject = (project: Project) => ({
+export const addProject = (project: Project, onboardingCompleted: boolean) => ({
```
Nit: Sorry I didn't think of this sooner, but how do you feel about changing this to `isOnboardingCompleted`, both here and everywhere else as well?

I like using `isX` for boolean variables as a quick way to tell whether the variable is a boolean. Plus then it'd be consistent with others (like `isVisible` below)

melanieseltzer replied:
Yeah I like how that sounds, kinda like a question with a yes answer, cause we'll only ever have 'true' as the value.

mathieudutour replied:
btw, not related to this PR but the type of `project` is wrong I believe. It should be `InternalProject` since it's in the reducer that it's transformed into a `Project`. Same for `importExistingProjectFinish`

joshwcomeau commented on src/sagas/import-project.saga.js:
```diff
@@ -71,6 +72,11 @@ export function* handleImportError(err: Error): Saga<void> {
 }
 
 export function* importProject({ path }: Action): Saga<void> {
+  const store = window.store.getState();
```
Ah, so `window.store` is meant to be used for debugging only. When we're within a saga, there are more idiomatic ways to get the state.

In this case, it'd be `yield select(selectorFunction)`.

You can use the selector function I mentioned in an earlier function, to do:

```js
const isOnboardingCompleted = yield select(getOnboardingCompleted);
```

melanieseltzer replied:
🤦‍♀️ Much nicer.

joshwcomeau commented on src/sagas/import-project.saga.test.js:
```diff
@@ -169,6 +175,10 @@ describe('import-project saga', () => {
       };
+
+      const getState = window.store.getState();
+      const onboardingCompleted = getState.onboardingStatus === 'done';
```
While it's less critical in tests, I think we should also avoid using `window.store` here.

In fact, for the purposes of this test, it's actually better if we hardcode it:

```js
expect(saga.next(jsonWithGuppy).value).toEqual(
  put(
          importExistingProjectFinish(
            'path/to/project',
            jsonWithGuppy,
            true // Use a hardcoded value
          )
  )
)
```

This way we can make sure the saga is passing along the right value, and not just whatever value it generates internally.

joshwcomeau commented:
> We might want to think about switching other instances of componentWillReceiveProps() as well?

Yeah, we'll eventually need to do this. We should probably make sure new code doesn't include any of the deprecated lifecycle methods.

In general we'd want to migrate to `getDerivedStateFromProps`, and not `componentDidUpdate`, since `componentDidUpdate` causes a double-render (since it happens _after_ the first update, if you change the state it does a whole new cycle). This isn't always super easy, since `getDerivedStateFromProps` gives you less information than `componentWillReceiveProps`, but yeah we should start moving in that direction.

melanieseltzer commented:
Pushed some updates addressing the feedback!

> In general we'd want to migrate to getDerivedStateFromProps, and not componentDidUpdate

Does that mean I should be using getDerivedStateFromProps here for the Application Menu, or did you mean use it for the other instances?

joshwcomeau commented:
> Does that mean I should be using getDerivedStateFromProps here for the Application Menu, or did you mean use it for the other instances?

In this case I don't think it matters :) the issue is when you have a `setState` call inside `componentDidUpdate`, since that means React is doing two render cycles (since `didUpdate` is called at the end of one, and then `setState` kicks off a new one). Because we're calling an Electron API here, it doesn't affect React rendering
